### ISSUE TITLE
Migrations and Installation Instructions Problem

### ISSUE THREAD

**Issue by objecttothis** [opensourcepos/opensourcepos#3646]
The [installation instructions](https://github.com/opensourcepos/opensourcepos/wiki/Getting-Started-installations) tell you to populate the empty database using the database.sql script.
![image](https://user-images.githubusercontent.com/17935339/216548066-06c0c9b3-78be-43db-8ada-63d4e209e4f8.png)

The problem with this is that the database.sql puts the database in a state that is after migration 20180225100000_upgrade_to_3_2_0

This means the three migrations below fail on new installs because they add columns that already exists.  If everyone was using MariaDB we could just add in IF NOT EXISTS after ADD COLUMN, but some are using MySQL which requires a stored procedure to do this.
![image](https://user-images.githubusercontent.com/17935339/216549136-e0f4b5b1-7177-4203-8404-aed9ec1bafee.png)

Unless I'm missing something, we have two options:
Option 1
We delete those three migrations from the repo, since new installs will already have their changes.

Option 2
We modify database.sql to bring it back to the structure just before running 20170501150000_upgrade_to_3_1_1.

Option 1 is easier, but it leaves anyone in the lurch trying to migrate from prior to OSPOS 3.2.0.  Option 2 retains backward compatibility.  Existing installs upgrading from prior to 3.0.2 will still need to manually run those SQL scripts before they could migrate from 3.0.2 using migrations, but they had to do that before anyway.

Unless anyone has any objections, I'm going to move forward with Option 2.  @opensourcepos/org thoughts?

**Comment by objecttothis** [comment]
To clarify, the problem is tables.sql.  database.sql is generated from that and constraints.sql

**Comment by objecttothis** [comment]
There is another problem.  20170501150000_upgrade_to_3_1_1 is the first migration and that adds columns to ospos_sales_suspended, but tables.sql doesn't create this table.  I'm finding references to altering that table back to 2.3.2_to_2.3.3.sql, but never a create statement.  I think we need to add the create table reference into the tables.sql at the 3.0.2 level so that those migrations and any others that are expecting the table don't fail.

**Comment by objecttothis** [comment]
OK, this is getting comical. I don't even have the ospos_sales_suspended table in my production instance.  Can anyone send me a table create script for this table?  If not, I'll figure out it's structure from the code, but it might be difficult because all I have are the ALTER statements

**Comment by objecttothis** [comment]
After doing some research, I found the create scripts in the database.sql from 2.2.2, but in 2.3 it was removed.  Any particular reason for removing it? As is the migrations just skip over all the alter statements because the table doesn't exist and on new installs the table never gets created.

**Comment by objecttothis** [comment]
OK, I found the commit that deleted these tables from the tables.sql (https://github.com/opensourcepos/opensourcepos/commit/86d3c6bf0492cf47bd8008515a2622c22f81a2f5).  @jekkos that was a commit that you made.  Without those tables, all the migration scripts that reference them don't make the changes due to the tables not existing on any installs made after that release.  Is it OK for me to add them back in to tables.sql, or is there a good reason to leave them out and do something different?

**Comment by jekkos** [comment]
@objecttothis the table was indeed removed at the time, we moved these properties to the sales table instead

**Comment by jekkos** [comment]
It could be that tables.sql were modified together with the migration files themselves.. so that might explain why we start the tables.sql 3 migrations further. There might have been two scenarios
* If one has an existing db, the 3 migrations were ran and version was determined to be older then the one that is preconfigured . Tables.sql is not in sync but this did not matter as it is not used
* One does not have a db yet, so we can start from database.sql and skip those 3 migrations.

**Comment by jekkos** [comment]
I think at that point the approach on the migrations wasn't fully clear, it was only after those three that we decided to freeze tables.sql. I'll think a bit how we can try to fix this properly.

**Comment by objecttothis** [comment]
OK, I think I see this better now.  In the 3.0.2_to_3.1.1.sql the tables are dropped
`DROP TABLE 'ospos_sales_suspended_payments', 'ospos_sales_suspended_items_taxes', 'ospos_sales_suspended_items', 'ospos_sales_suspended';`
Above that, however, in the same script above the tables are referenced several times which assumes that the table exists.

**Comment by objecttothis** [comment]
> I think at that point the approach on the migrations wasn't fully clear, it was only after those three that we decided to freeze tables.sql. I'll think a bit how we can try to fix this properly.

Agreed that tables.sql needs to be frozen, but it should be frozen at a state just prior to the first automated migration, otherwise it fails.

**Comment by objecttothis** [comment]
> It could be that tables.sql were modified together with the migration files themselves.. so that might explain why we start the tables.sql 3 migrations further. There might have been two scenarios
> 
>     * If one has an existing db, the 3 migrations were ran and version was determined to be older then the one that is preconfigured . Tables.sql is not in sync but this did not matter as it is not used
> 
>     * One does not have a db yet, so we can start from database.sql and skip those 3 migrations.

OK, this is helping me think through it a little.  Here are the scenarios I think users will encounter if we delete those three migrations

- Installing a fresh instance: No problems because database.sql has those changes baked in.
- Upgrading from 3.0.2 or 3.1.1: They aren't using database.sql, but changes that happened in the three deleted migrations are never made.  That's not tenable.
- Upgrading from 3.2.0 or later: No problems because changes were already made and nothing is missed.
- Upgrading from prior to 3.0.2: They run all the SQL scripts prior to 3.0.2_to_3.1.1, but can't make the changes in those three deleted migrations.

Since there are two scenarios in which users would not have an upgrade path, I think we should continue to freeze tables.sql and constraints.sql but at the point of 3.0.2, not 3.2.0. In this scenario no deleting of migrations needs to happen and everyone has an upgrade path.

**Comment by objecttothis** [comment]
@jekkos I just did a quick test by creating a clean database from the 3.0.2 version of tables.sql and constraints.sql and the migrations are running on the CI4 migrator without the errors caused by columns already existing.

**Comment by objecttothis** [comment]
I just pushed a commit that makes good progress on fixing problems in code that was working in CI3 for migrations.  All the migrations up to 20210422000000_database_optimizations run using the UI, but php spark migrate fails on any migrations that use the in-memory appconfig data because that doesn't load for the command line.  If we want to be able to run migrations from spark, then we will need to modify these migrations to pull the appconfig values straight from the database rather than in memory.

If anyone wants to pick this up over the weekend, feel free.  On a clean database created from the database.sql in OSPOS 3.0.2 navigate to /public/ and login with the default username and password.  It should run migration all the way up to 20210422000000_database_optimizations and then error out showing you the stack trace.  The development error display is much more verbose and helpful.

**Comment by objecttothis** [comment]
The migrations are fixed now in the CI4 branch.  Unless there is a compelling reason not to, I've reverted the tables.sql and constraints.sql to the 3.0.2 version, because with the current version the first three automated migrations fail since they address columns that were removed from tables.sql and constraints.sql in 3.2.0.  We should freeze these two files at their current version so that all future migrations work properly.
